feat: enable code fence capabilities on mobile app.

Jonathansoufer opened the pull request:
## **Description**

This PR implements code fence feature to allow flask and main code co-live in the same branch yet delivered accordantly  with build type, meaning flask code will only be included in flask build and not main.

[Here](https://github.com/MetaMask/metamask-extension/blob/a0e8ee14af00f9793349206909c44d7b68985c41/development/build/transforms/remove-fenced-code.js#L88) is the version of code fencing implemented on extension.

## **Related issues**

Fixes: https://github.com/MetaMask/accounts-planning/issues/112

## **Manual testing steps**

Add code fence directive (BEGIN and END) on any code block and run the build.

```
              {
               ///: BEGIN:ONLY_INCLUDE_IN(flask)
              }

              <View style={styles.foxWrapper}>
                {Device.isAndroid() ? (
                  <Image
                    source={require('../../../images/fox.png')}
                    style={styles.image}
                    resizeMethod={'auto'}
                  />
                ) : (
                  <AnimatedFox bgColor={colors.background.default} />
                )}
              </View>
              {
                ///: END:ONLY_INCLUDE_IN
              }
``` 

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

## **Next Steps**
- [ ] Adjust `android/app/build.gradle` and add new iOS configurations/schemas for FLASK build.
- [ ]  Add env variable to distinguish builds.

Commit:
new: add turning comments off so that babel wont includes comments in non-prod builds

Commit:
new: fence code core transformation based on metro bundler

Commit:
new: adds basic set of tests

Commit:
new: installs correct version of metro-react-native-babel-transformer

Commit:
refactor: removes leftover semicomma

Commit:
refactor: polish guard brackets

owencraston commented:
we also need to ensure that we run a lint step after we build both main and flask to ensure that there are no unused variables and that each `///: END:ONLY_INCLUDE_IN` has a corresponding `///: END:ONLY_INCLUDE_IN`

Commit:
refactor: polish tests

Commit:
refactor: polish on prettier reqs

Commit:
refactor: polish on prettier reqs

sethkfman commented on metro.transform.js:
```diff
@@ -0,0 +1,35 @@
+  if (['.js', '.cjs', '.mjs', '.ts'].includes(path.extname(filename))) {
+    const [processedSource, didModify] = removeFencedCode(
+      filename,
+      { all: new Set(['flask', 'snaps']), active: new Set([]) },
```
Can we break this array out to a variable at the top of the file to make it easier to configure/add/reference in a ReadMe?

Jonathansoufer replied:
Got it. Done.

sethkfman commented:
@Jonathansoufer The only thing I would add is a bit of documentation in the README.md with a short description of code fencing and how to use it. If you wrote something up in notion bring it into the project.

owencraston commented:
@Jonathansoufer we can probably add a similar readme to [this one](https://github.com/MetaMask/metamask-extension/blob/a0e8ee14af00f9793349206909c44d7b68985c41/development/build/transforms/README.md)

Cal-L commented on metro.config.js:
```diff
@@ -33,7 +33,6 @@ module.exports = (async () => {
     resolver: {
       assetExts: assetExts.filter((ext) => ext !== 'svg'),
       sourceExts: [...sourceExts, 'svg', 'cjs'],
-      resolverMainFields: ['sbmodern', 'react-native', 'browser', 'main'],
```
What was the reason for this change?

Cal-L commented on remove-fenced-code.js:
```diff
@@ -0,0 +1,473 @@
+ * @returns {[string, modified]} A tuple of the post-transform file contents and
+ * a boolean indicating whether they were modified.
+ */
+function removeFencedCode(filePath, features, fileContent) {
```
Instead of passing in both `filePath` and `fileContent`, does it make sense to only pass `filePath` and read it within here?

Jonathansoufer replied:
Not in this case. Since we're using the same method signature on both mobile and web and taking advantage that the default bundler signature (https://facebook.github.io/metro/docs/getting-started/#javascript-transformer) already gives both src/filename, we don't need to redo it inside the removeFencedCode for the sake of removing one param. Thanks for bringing this to light so we can explore/explain more.

Cal-L commented on remove-fenced-code.test.js:
```diff
@@ -0,0 +1,56 @@
+    }).toThrow(/Invalid fence line in file/);
+  });
+
+  it('should throw an error if a fence is malformed', () => {
```
Why is this throwing a different error vs the test above?

Jonathansoufer replied:
All UTs are passing now. Could you point which discrepancies you see on these, if any still?

owencraston replied:
The function was not actually throwing different errors, it was simply throwing one big error and there were two tests checking each component of the error message. I modified this so that we combined this test into one to be more clear. I also added more specific tests for the following cases...

- invalid fence param
- invalid fence line
- fences with multiple params
- files with multiple fences of different types
- expected output when the activeFeature IS set to `flask`

Cal-L reviewed (commented):
Left some comments

Commit:
refactor: using array as const  making it easier to configure/add/reference in a ReadMe

Commit:
refactor: adding back wrongly removed resolver

Commit:
refactor: fixed should remove fenced code test case

Commit:
refactor: bootstrap README for mobile code fence

Commit:
refactor: add README and new folder structure

Commit:
refactor: add transforms folder structure

Commit:
refactor: add availableFeatures const

owencraston committed:
run yarn install to dedupe

owencraston commented on metro.transform.js:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  function getBuildTypeFeatures() {
+    switch (process.env.BUILD_TYPE) {
```
The flask build type will be created in [this PR](https://github.com/MetaMask/metamask-mobile/pull/7837)

owencraston committed:
slit up tests and add invalid param test

owencraston committed:
test multiple code fences in one file

owencraston commented:
I added a few more specific tests so that the expected results were more clear and they are all passing.
<img width="975" alt="Screenshot 2023-11-16 at 3 35 31 PM" src="https://github.com/MetaMask/metamask-
mobile/assets/22918444/75c83add-444c-4b13-b593-c5a30d9d8327">

owencraston committed:
fix Runaway Regular Expressions: Catastrophic Backtracking

- see: https://www.regular-expressions.info/catastrophic.html

owencraston commented on app/transforms/remove-fenced-code.js:
```diff
@@ -0,0 +1,473 @@
+// newlines. We except newlines to avoid eating blank lines preceding a fenced
+// line.
+// Double-negative RegEx credit: https://stackoverflow.com/a/3469155
+const linesWithFenceRegex = /^[^\S\r\n]*\/\/\/:[^\r\n]*$/gm;
```
I had to modify this regex to fix [this security issue](https://sonarcloud.io/organizations/consensys/rules?open=javascript%3AS5852&rule_key=javascript%3AS5852&tab=how_to_fix). All the tests are still passing

owencraston committed:
introduce METAMASK_BUILD_TYPE

- METAMASK_BUILD_TYPE can either be 'main' or 'flask'
- All instances of METAMASK_BUILD_TYPE should be set to 'main' for now.
  Flask will be introduced later.

owencraston commented on metro.transform.js:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  function getBuildTypeFeatures() {
+    const buildType = process.env.METAMASK_BUILD_TYPE ?? 'main';
```
the default value is `main`. This was needed to fix CI which was failing because the variable BUILD_TYPE did not have a valid value. I changed this variable to be `METAMASK_BUILD_TYPE` which was the agreed upon pattern from [this PR](https://github.com/MetaMask/metamask-mobile/pull/6645/files). For now all values of `METAMASK_BUILD_TYPE` should be set to `main`. We will introduce the `flask` `METAMASK_BUILD_TYPE` in [this following PR](https://github.com/MetaMask/metamask-mobile/pull/7837).

Jonathansoufer replied:
Good one.

owencraston committed:
correct error message

owencraston committed:
add METAMASK_BUILD_TYPE to detox

owencraston committed:
export METAMASK_BUILD_TYPE

owencraston commented:
I re ran the e2e: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/883c5038-5267-461a-9a4b-749e939c56a4

Cal-L reviewed (approved):
LGTM

owencraston commented:
the pr regression e2e also passed: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/4967eb2b-de95-42d7-af15-7041d7f07647

chrisleewilcox committed:
Merge branch 'main' into feat/add-flask-code-fencing
